helix-term/src/ui/picker.rs

```diff
@@ -113,17 +162,46 @@ impl<T: 'static> Component for FilePicker<T> {
             horizontal: 1,
         };
         let inner = inner.inner(&margin);
-
         block.render(preview_area, surface);
 
-        if let Some((doc, line)) = self.current_file(cx.editor).and_then(|(path, range)| {
-            cx.editor
-                .document_by_path(&path)
-                .or_else(|| self.preview_cache.get(&path))
-                .zip(Some(range))
-        }) {
+        if let Some((path, range)) = self.current_file(cx.editor) {
+            let doc = match self.get_preview(path, cx.editor) {
```

Comment:
This whole `match` statement can be replaced with

```rust
            let preview = self.get_preview(path, cx.editor);
            if !preview.document().is_some() {
                let alt_text = preview.placeholder();
                surface.set_stringn(inner.x, inner.y, alt_text, inner.width as usize, text_style);
                return;
            }
            let doc = preview.document().unwrap();
```

provided that we define a few convenience methods on the `Preview` enum:

```rust
impl Preview<'_, '_> {
    pub fn document(&self) -> Option<&Document> {
        match *self {
            Self::EditorDocument(doc) => Some(doc),
            Self::Cached(CachedPreview::Document(doc)) => Some(doc),
            _ => None,
        }
    }

    pub fn placeholder(&self) -> &str {
        match *self {
            Self::EditorDocument(_) => "<File preview>",
            Self::Cached(preview) => match preview {
                CachedPreview::Document(_) => "<File preview>",
                CachedPreview::Binary => "<Binary file>",
                CachedPreview::LargeFile => "<File too large to preview>",
                CachedPreview::NotFound => "<File not found>",
            },
        }
    }
}
```

Reply:
I think we should not have unwrap.

```rust
            let preview = self.get_preview(path, cx.editor);
            if !preview.document().is_some() {
                let alt_text = preview.placeholder();
                surface.set_stringn(inner.x, inner.y, alt_text, inner.width as usize, text_style);
                return;
            }
            let doc = preview.document().unwrap();
```
So I write this.

```rust
            let preview = self.get_preview(path, cx.editor);
            let doc = match preview.document() {
                   Some(doc) => doc,
                   None => {
                       let alt_text = preview.placeholder();
                       surface.set_stringn(inner.x, inner.y, alt_text, inner.width as usize, text_style);
                       return;
                   }
             }
```